### PR TITLE
Add basis of S3 bucket creation

### PR DESCRIPTION
**Pull request by ciaranevans**
* Adds secret encrypted parameters to SSM
* Maintains infra state in S3
* Contains scripts to encrypt/decrypt secrets - #encrypted #decrypted

PBI:146656

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,7 @@ build/
 *.iml
 
 *.swp
+
+*.secret.*
+
+.terraform/
```
Comment (eganjs) [inline .gitignore]: add new line at end of file

File: src/main/terraform/infra_deployment/main.tf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+resource "aws_s3_bucket" "ais_bucket" {
+  bucket = "${data.aws_ssm_parameter.ais_bucket_name.value}"
+  acl = "private"
+}
+
+resource "aws_s3_bucket" "heatmap_bucket" {
+  bucket = "${data.aws_ssm_parameter.heatmap_bucket_name.value}"
+  acl = "private"
+}
```
Comment (eganjs) [inline src/main/terraform/infra_deployment/main.tf]: add new line at end of file

File: src/main/terraform/parameter_deployment/encrypt_parameters.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+openssl enc -aes-256-cbc -k $AIS_TO_HEATMAPS_PARAMETERS_PASSWORD -in parameters.secret.json -out parameters.enc.json
+
+cat <<EOF
+     __ __                                   __           __
+  __/ // /____  ____  ____________  ______  / /____  ____/ /
+ /_  _  __/ _ \/ __ \/ ___/ ___/ / / / __ \/ __/ _ \/ __  /
+/_  _  __/  __/ / / / /__/ /  / /_/ / /_/ / /_/  __/ /_/ /
+ /_//_/  \___/_/ /_/\___/_/   \__, / .___/\__/\___/\__,_/
+                             /____/_/
+EOF
```
Comment (eganjs) [inline src/main/terraform/parameter_deployment/encrypt_parameters.sh]: add new line at end of file

File: src/main/terraform/parameter_deployment/main.tf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  name = "/buckets/ais_bucket"
+  type = "SecureString"
+  value = "${data.external.secrets.result["ais_bucket"]}"
+}
+
+resource "aws_ssm_parameter" "heatmap_bucket" {
+  name = "/buckets/heatmap_bucket"
+  type = "SecureString"
+  value = "${data.external.secrets.result["heatmap_bucket"]}"
+}
```
Comment (eganjs) [inline src/main/terraform/parameter_deployment/main.tf]: add new line at end of file

File: src/main/terraform/infra_deployment/main.tf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+terraform {
+  backend "s3" {
+    bucket = "ais-to-raster-terra-state"
+    key    = "infra-state"
+    region = "eu-west-2"
+  }
+}
+
+data "aws_ssm_parameter" "jobs_bucket_name" {
```
Comment (eganjs) [inline src/main/terraform/infra_deployment/main.tf]: Consider moving these data statements into a variables tf file

File: src/main/terraform/parameter_deployment/main.tf [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  backend "s3" {
+    bucket = "ais-to-raster-terra-state"
+    key    = "parameter-state"
+    region = "eu-west-2"
+  }
+}
+
+variable "PASSWORD" {}
+
+data "external" "secrets" {
```
Comment (eganjs) [inline src/main/terraform/parameter_deployment/main.tf]: Consider moving these data statements into a variables tf file

File: .gitignore (hunk 1 above)
Comment (ciaranevans) [reply .gitignore]: done

File: src/main/terraform/infra_deployment/main.tf (hunk 2 above)
Comment (ciaranevans) [reply src/main/terraform/infra_deployment/main.tf]: done

File: src/main/terraform/parameter_deployment/encrypt_parameters.sh (hunk 3 above)
Comment (ciaranevans) [reply src/main/terraform/parameter_deployment/encrypt_parameters.sh]: done

File: src/main/terraform/parameter_deployment/main.tf (hunk 4 above)
Comment (ciaranevans) [reply src/main/terraform/parameter_deployment/main.tf]: done

File: src/main/terraform/infra_deployment/main.tf (hunk 5 above)
Comment (ciaranevans) [reply src/main/terraform/infra_deployment/main.tf]: Considered. Will move if these become too big.

File: src/main/terraform/parameter_deployment/main.tf (hunk 6 above)
Comment (ciaranevans) [reply src/main/terraform/parameter_deployment/main.tf]: Ditto

File: src/main/terraform/infra_deployment/vars.tf [hunk 7]
```diff
@@ -0,0 +1,8 @@
+variable "PASSWORD" {}
+
+data "external" "secrets" {
+  program = [
+    "openssl", "enc", "-aes-256-cbc", "-d", "-k", "${var.PASSWORD}", "-in",
+    "${path.module}/parameters.enc.json"
+  ]
+}
```
Comment (eganjs) [inline src/main/terraform/infra_deployment/vars.tf]: Add new line at end of file

Comment (ciaranevans) [reply src/main/terraform/infra_deployment/vars.tf]: done

**Pull request merged**